### ISSUE TITLE
Socket6 bug in 5.36 and 5.38

### ISSUE THREAD

**Issue by sidney** [StrawberryPerl/Perl-Dist-Strawberry#156]
This appears to be a Strawberry bug rather than a Socket6 bug because 1) It works in Strawberry Perl 5.32 but not in 5.36 and 5.38; 2) It works on Linux in all versions of perl; 3) See Issue #72 and all the patches to Socket6 that were put in to get it to build in Strawberry 5.36.

One-liner that demonstrates bug. It works in Strawberry Perl 5.32 and also on other platforms in recent perl versions.

`perl -MSocket6 -e "print length(Socket6::inet_pton(Socket6::AF_INET6(), '::1:2:3:4:5:6:7')) || 'error undefined, should be 16';"`

Some addresses I tried, such as `'::2:3:4:5:6:7'` work, some other valid ipv6 addresses don't.

Note that NetAddr::IP now fails its tests in Strawberry Perl 5.36 and 5.38 because it has a test case that calls exactly this.

**Comment by shawnlaffan** [comment]
And FWIW, NetAddr-IP-4.079 builds and passes all its tests when using SP 5.36.1.  This is for both the XS and pure perl versions.

**Comment by sisyphus** [comment]
> I cannot reproduce on my local machine when using SP 5.36 or 5.38 portable editions.

I cannot reproduce it on my SP builds on Windows 11 either.
@sidney, what version of Windows are you running ?

**Comment by sidney** [comment]
This is great news, it means there is something different about my Windows setup that I can try to narrow down. Right now I am running it on a Windows 10 Pro version 22H2 OS build 19045.3693 as a VM under VMware Fusion on a Mac host. I thought I had the same result on Windows 11, but now that you have shown that it can work, I need to more carefully run tests. In all cases I installed Strawberry Perl from the msi, not the portable edition. I will try out different Windows, network configurations, and with portable editions, and come back with a more definitive list of what does and does not work.

**Comment by sidney** [comment]
It does work for me too on Windows 11, and not just with portable edition.

I apologize for my mistake... After going back and forth between Windows 10 and 11 tracking down different things I thought I had run the same tests regarding Socket6 and NetAddr::IP on both, but apparently the final tests isolating the problem were just Windows 10. I will come back after trying again with variations on Windows 10.

**Comment by shawnlaffan** [comment]
I can also reproduce using a Windows 10 box.  This is the same machine used to build SP 5.38 and 5.36.

**Comment by sidney** [comment]
@shawnlaffan If you can reproduce on Windows 10 I won't try to experiment with variations on that platform. Are you able to put together a small test in C to see how the Windows 10 native inet_pton (https://learn.microsoft.com/en-us/windows/win32/api/ws2tcpip/nf-ws2tcpip-inet_pton) translates an ipv6 address '::1:2:3:4:5:6:7' ?

**Comment by shawnlaffan** [comment]
> Are you able to put together a small test in C to see how the Windows 10 native inet_pton translates an ipv6 address '::1:2:3:4:5:6:7' ?

Not easily.  

Hopefully someone else can chime in with some code.

**Comment by sidney** [comment]
I asked chatgpt to write the test code and it did!

I'm attaching Windows and Linux/unix versions of a small test using the ANSI C function inet_pton.The program prompts for the ipv6 address string to use. Here are results I got in Windows 10, Windows 11, and macOS

So it looks like a bug in the windows 10 inet_pton implementation. 

Windows 10

```
Enter an IPv6 address: ::1:2:3:4:5:6:7
Invalid IPv6 address: ::1:2:3:4:5:6:7
```

```
Enter an IPv6 address: ::1:2:3:4:5:6
Length of the resulting data: 16

```
macOS

```
Enter an IPv6 address: ::1:2:3:4:5:6:7
Length of the resulting data: 16
```

Windows 11

```
Enter an IPv6 address: ::1:2:3:4:5:6:7
Length of the resulting data: 16

```
To build the Windows version in a Visual Studio command shell
`  cl ptontestwindows.c /EHsc /link Ws2_32.lib`

To build the linux/unix version using gcc

`   gcc ptontestlinux.c -o ptontestlinux`

[ptontest.tgz](https://github.com/StrawberryPerl/Perl-Dist-Strawberry/files/13519426/ptontest.tgz)

**Comment by shawnlaffan** [comment]
Thanks for the code and results.  

This looks relevant:  https://github.com/msys2/MINGW-packages/issues/6191#issuecomment-586067783

Searching for ```_WINNT``` in the code points to the HelpCA.dll_src and HelpCA_v2.dll_src directories, e.g. https://github.com/StrawberryPerl/Perl-Dist-Strawberry/blob/427b95661366260b251015a2fb90fd2745aa0cbb/HelperCA.dll_src/src/stdafx.h#L15

However, those seem only to be used for the MSI files, specifically the installers and not the main SP compilation.

**Comment by shawnlaffan** [comment]
See also #77

**Comment by sidney** [comment]
Is this relevant? I get confused reading it, but I'm linking it here anyway for anyone who finds it less confusing in case it helps.  https://rt.cpan.org/Ticket/Display.html?id=140028

**Comment by sidney** [comment]
Let me see if I have this right: The minimum version of Windows to run Strawberry is Vista/8 which corresponds to _WIN32_WINNT set to 0x0600. The patches have to do with getting Socket6 XS to build correctly with the 0x0600 target. In the Socket6 code there are #ifdef lines that decide whether the inet_pton function exists in the win32 api version that is being targeted, or if it should use its own implementation instead. Because the win32 api as of version 0x0600 does implement inet_pton, the XS build of Socket6 uses it. However, the Windows 10 implementation is buggy, failing when passed the ipv6 address ::1:2:3:4:5:6:7 which according to the relevant RFC is equivalent to 0:1:2:3:4:5:6:7 and  is recognized by inet_pton in Windows 11 and various linux and unix platform APIs.

If I'm getting this right, then the patch to fix this would be in Socket6 wherever it has #ifdef deciding that the win32 api version is one that has native inet_pton, and patch it to always use the non-native function. This will be necessary as long as Strawberry supports Windows 10.

**Comment by sidney** [comment]
I don't know where the right place to put a patch is, but it looks like the highest level place that should produce the minimum necessary change is in Socket6 Makefile.PL in `sub configure_mswin32`

to insert two lines to change
```
        } elsif (/HAVE_INET_(PTON|NTOP)/) {
```
to
```
        } elsif (/HAVE_INET_PTON/) {
             	    print(OUT "  #undef HAVE_INET_PTON\n");
        } elsif (/HAVE_INET_(PTON|NTOP)/) {

```
so it unconditionally uses the inet_pton.c that is in the source code instead of the win32 api version.

Is this something you can try building to see if it works?

**Comment by shawnlaffan** [comment]
That works for me on both the Windows 10 and 11 boxes.  

The place to report it is to the RT ticket: https://rt.cpan.org/Public/Bug/Display.html?id=140028

There has not been a release of Socket6 for a while so unless there are objections I will prep a new patched version for the next SP build (we currently distribute a patched Socket6-0.29_01).

**Comment by sidney** [comment]
I'm commenting on the RT ticket now

**Comment by shawnlaffan** [comment]
@sidney - Strawberry Perl 5.38.2.1 includes a patched version of Socket6 based on the above discussion.  

Can you give it a test to see if it works for you?

https://github.com/StrawberryPerl/Perl-Dist-Strawberry/releases/tag/SP_53821_64bit

**Comment by sidney** [comment]
@shawnlaffan - Not the best news... It does give the correct answers for Socket6::inet_pton, and NetAddr::IP successfully passes its tests and installs on Windows 10, which fixes the bug I reported, so that much is better.

However, I then tried `cpanm -v --test-only` my current development build of Mail::SpamAssassin, which requires NetAddr::IP and is the reason I discovered this problem. The results:

Windows 10,  perl 5.38.0 :  could not test Mail::SpamAssassin because the bug doesn't let me install NetAddr::IP
Windows 11, perl 5.38.0:  all required modules installed and all Mail::SpamAssassin tests passed

Windows 10 and 11, perl 5.38.2 : all required modules installed, but the one test that runs a local nameserver fails. I have not yet tracked down exactly what fails to be able to tell you exactly what the failure is or even what module it is in. I only have the hint that it is the only test that tries to run and query a local nameserver, and the test exits prematurely with an error code.

I'll investigate some more to see if I can get a more specific bug to report.

**Comment by sidney** [comment]
Ok, this is probably not your bug. 

The above was run with Net::DNS 1.41 installed. That was a version recently released to fix a bug I reported in https://rt.cpan.org/Ticket/Display.html?id=150550 

with a small test case I pasted into comment https://rt.cpan.org/Transaction/Display.html?id=2618099

As I posted there, the patch I suggested is a correct fix for a real problem, but is not sufficient to get the test case to run correctly. It prevents the test case from erroring out, but it still ends up hanging. The test failure I now see in SpamAssassin when using Net::DNS 1.41 is the result of that hang.

The bottom line is that in versions of Net::DNS 1.38 and greater Net::DNS::Nameserver does not work on Windows, but it fails with a hang instead of an error crash in version 1.41.

I'll bring that up in the rt ticket for Net::DNS, although if you can figure out how to get that test case to work instead of hanging, that will be great. I'll be investigating it more in any case.

**Comment by shawnlaffan** [comment]
Thanks @sidney.  I see there is a new dev release of Mail::SpamAssassin on CPAN (https://metacpan.org/release/SIDNEY/Mail-SpamAssassin-4.0.1-pre1i-TRIAL/changes) so assume things are in train on that front.  

I think this issue is now closable but I'll leave it open for a few days (maybe more) in case something else comes up that is closely related.

**Comment by shawnlaffan** [comment]
I'll close this issue now but it can be re-opened if needed.
